File: IPWorks Samples/Certificate Manager/cpp/certmgr.cpp
/*
 * IPWorks 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworks.h"
#define LINE_LEN 100

/**
 * README
 * If the options are selected sequentially no certificates or keys
 * are required everything used will be created by the demo.
 **/


class MyCertMgr : public CertMgr
{
private :
	static int serialNumber;
	char subjectCA[1024];
	char subjectCert[1024];
public:
	MyCertMgr()
	{
		Config("CertValidityTime=10");
		Config("CertKeyLength=1024");
		strncpy(subjectCA,"C=US,ST=North Carolina,L=Durham,O=/n Software,OU=CA",1023);
		strncpy(subjectCert,"C=US,ST=North Carolina,L=Durham,O=/n Software,OU=Development",1023);
	}

	virtual int FireError(CertMgrErrorEventParams *e)
	{
		printf("Error Code :%d\n",e->ErrorCode);
		printf("Error Description :%s\n",e->Description);
		return 0;
	}

	void exportToFile()
	{
		char filename[1024];
		char password[1024];
		printf("Export certificate,key to file : [store.pfx] :");
		fgets(filename,1024,stdin);
		if( strlen(filename) == 1 ) strcpy(filename,"store.pfx");
		else filename[strlen(filename)-1] = '\0';
		printf("Password for private key encryption :");
		fgets(password,1024,stdin);
		password[strlen(password)-1] = '\0';
		if ( !ExportCertificate(filename,password))
		{
			printf("Certificate & key exported successfully.\n");
		}
		else
		{
			printf("Error :%s\n",GetLastError());
		}
	}

	void setStore()
	{
		SetCertStoreType(CST_PFXFILE);
		char filename[1024];
		char password[1024];
		printf("Store file : [store.pfx] :");
		fgets(filename,1024,stdin);
		if( strlen(filename) == 1 ) strcpy(filename,"store.pfx");
		else filename[strlen(filename)-1] = '\0';
		printf("Store password :");
		fgets(password,1024,stdin);
		password[strlen(password)-1] = '\0';
		SetCertStorePassword(password);
		if ( !SetCertStore(filename,strlen(filename)))
		{
			printf("Store set successfully.\n");
		}
		else
		{
			printf("Error :%s\n",GetLastError());
		}
	}

	void readCert()
	{
		char filename[1024];
		printf("Read Certificate file : [cert.pem] :");
		fgets(filename,1024,stdin);
		if( strlen(filename) == 1 ) strcpy(filename,"cert.pem");
		else filename[strlen(filename)-1] = '\0';
		if(! ReadCertificate(filename))
		{
			printf("Certificate read successfully.\n");
		}
		else
		{
			printf("Error :%s\n",GetLastError());
		}
	}

	void printCert()
	{
		printf("**** Certificate Details *****\n");
		printf("Certificate Effective Date :%s\n",GetCertEffectiveDate());
		printf("Certificate Expiration Date :%s\n",GetCertExpirationDate());
		printf("Certificate Issuer :%s\n",GetCertIssuer());
		printf("Certificate Subject :%s\n",GetCertSubject());
		printf("Certificate Usage :%s\n",GetCertUsage());
		printf("Certificate Public Key :%s\n",GetCertPublicKey());
		printf("Certificate Public Key Algorithm :%s\n",GetCertPublicKeyAlgorithm());
		printf("Certificate Public Key Length :%d\n",GetCertPublicKeyLength());
		printf("Certificate Serial Number :%s\n",GetCertSerialNumber());
		printf("Certificate Signature Algorithm :%s\n",GetCertSignatureAlgorithm());
		printf("Certificate Thumbprint SHA1 :%s\n",GetCertThumbprintSHA1());
		printf("Certificate Thumbprint MD5 :%s\n",GetCertThumbprintMD5());
	}

	void saveCert()
	{
		char filename[1024];
		printf("Save certificate to file : [cert.pem] :");
		fgets(filename,1024,stdin);
		if( strlen(filename) == 1 ) strcpy(filename,"cert.pem");
		else filename[strlen(filename)-1] = '\0';
		if( !SaveCertificate(filename))
		{
			printf("Certificate saved successfully.\n");
		}
		else
		{
			printf("Error :%s\n",GetLastError());
		}
	}

	void createCert(int selfsigned)
	{
		char *store = NULL;
		int length = 0;
		if(selfsigned)
		{
			if (!CreateCertificate(subjectCA,-1) )
			{
				printf("Certificate created successfully.\n");
			}
			else
			{
				printf("Error :%s\n",GetLastError());
			}
		}
		else
		{
			GetCertStore(store,length);
			if( !store )
			{
				printf("There is no certificate in the store, try a self signed certificate.\n");
				return;
			}
			if ( !IssueCertificate(subjectCert,serialNumber++) )
			{
				printf("Certificate created successfully.\n");
			}
			else
			{
				printf("Error :%s\n",GetLastError());
			}
		}
	}
};

int MyCertMgr::serialNumber = 1;

char get_selection()
{
	char choice[20];
	do
	{
		printf("\n********************************\n");
		printf("[a] Create a self signed certificate.\n");
		printf("[b] Export the certificate and the private key to a file.\n");
		printf("[c] Assign a certificate to the store.\n");
		printf("[d] Create a certificate signed with the certificate in the store.\n");
		printf("[e] Save the certificate to a file.\n");
		printf("[f] Read certificate from a file.\n");
		printf("[g] Print certificate details.\n");
		printf("[q] Quit.\n");
		printf("What would you like to do? ");
		fgets(choice,20,stdin);
	}
	while (strlen(choice) != 2);
	return choice[0];
}

int main(int argc, char **argv)
{
	MyCertMgr certmgr;
	int quit = 0;
	do
	{
		switch(get_selection())
		{
			case 'a':
			case 'A':
				certmgr.createCert(1);
				break;

			case 'b':
			case 'B':
				certmgr.exportToFile();
				break;

			case 'c':
			case 'C':
				certmgr.setStore();
				break;

			case 'd':
			case 'D':
				certmgr.createCert(0);
				break;

			case 'f':
			case 'F':
				certmgr.readCert();
				break;

			case 'G':
			case 'g':
				certmgr.printCert();
				break;

			case 'e':
			case 'E':
				certmgr.saveCert();
				break;

			case 'q':
			case 'Q':
				quit = 1;
				break;

			default:
				printf("Invalid option. Please select a letter from the menu below.\n");
		}
	}
	while(!quit);
	return 0;
}

File: IPWorks Samples/Certificate Manager/java/certmgr.java
/*
 * IPWorks 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworks.*;

public class certmgr {
    
    static int i = 0;
    
    public static void main(String[] args) {
    	
    	if(args.length!=2) {
			
			System.out.println("usage: certmgr filename password");
			System.out.println("");
			System.out.println("  filename the path to the file containing certificates and optional private keys");
			System.out.println("  password the password for the certificate store file.If test file is used, set the password to \"test\"");
			System.out.println("\r\nExample: certmgr test.pfx test");
		
		} else {
			
			CertMgr certmgr1= new CertMgr();
	    	
	    	try {
	            	    		
	            certmgr1.addCertMgrEventListener(new DefaultCertMgrEventListener(){
	            	public void certList(CertMgrCertListEvent e) {
	            		i++;
	                    System.out.println(i + ". " + e.certSubject);
	                }	            	
	            });	           
	            certmgr1.setCertStoreType(certmgr1.cstJKSFile); //user java key store (JKS file)
	            certmgr1.setCertStore(args[0]);
	            certmgr1.setCertStorePassword(args[1]); //If test file (found in demo folder) is used, set the password to "test"
	            certmgr1.listStoreCertificates();
	        } catch(Exception ex){
	            System.out.println(ex.getMessage());
	        }
	        System.out.println("Exited.");
		}
    	
    }        
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksException) {
      System.out.print(" (" + ((IPWorksException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}



